src/github-app.ts

```diff
@@ -1,19 +1,120 @@
+import cacheManager from 'cache-manager'
 import jwt from 'jsonwebtoken'
+import { Context, WebhookEvent } from './context'
+import { GitHubAPI } from './github'
+import { logger } from './logger'
+import { LoggerWithTarget, wrapLogger } from './wrap-logger'
 
-export const createApp = (options: AppOptions) => {
-  return () => {
+const cache = cacheManager.caching({
+  store: 'memory',
+  ttl: 60 * 60 // 1 hour
+})
+
+// Some events can't get an authenticated client (#382):
+function isUnauthenticatedEvent (event: WebhookEvent) {
+  return !event.payload.installation ||
+    (event.event === 'installation' && event.payload.action === 'deleted')
+}
+
+export interface Options {
+  id: number
+  cert: string
+}
+
+export class GitHubApp {
+  public log: LoggerWithTarget
+  public id: number
+  public cert: string
+
+  /**
+   * @param id - ID of the GitHub App
+   * @param cert - The private key of the GitHub App
+   */
+  constructor (id: number, cert: string) {
+    this.id = id
+    this.cert = cert
+    this.log = wrapLogger(logger, logger)
+  }
+
+  /**
+   * Create a new JWT, which is used to [authenticate as a GitHub
+   * App](https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/#authenticating-as-a-github-app)
+   */
+  public jwt () {
     const payload = {
       exp: Math.floor(Date.now() / 1000) + 60,  // JWT expiration time
       iat: Math.floor(Date.now() / 1000),       // Issued at time
-      iss: options.id                           // GitHub App ID
+      iss: this.id                              // GitHub App ID
     }
 
     // Sign with RSA SHA256
-    return jwt.sign(payload, options.cert, { algorithm: 'RS256' })
+    return jwt.sign(payload, this.cert, { algorithm: 'RS256' })
   }
-}
 
-export interface AppOptions {
-  id: number
-  cert: string
+  public async createContext (event: WebhookEvent) {
+    const log = this.log.child({ name: 'event', id: event.id })
+
+    let github
+
+    if (isUnauthenticatedEvent(event)) {
+      github = await this.auth()
+      log.debug('`context.github` is unauthenticated. See https://probot.github.io/docs/github-api/#unauthenticated-events')
+    } else {
+      github = await this.auth(event.payload.installation.id, log)
+    }
+
+    return new Context(event, github, log)
+  }
+
+  /**
+   * Authenticate and get a GitHub client that can be used to make API calls.
```

Comment:
I know this is just copy pasted form the old but maybe here would be a good place to clarify that GitHub client = authenticated as an installation: https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/#authenticating-as-an-installation